### PR TITLE
Adding ancient subworkflow

### PR DESCRIPTION
**Pull request by maxibor**
This pull request adds a subworkflow specific to ancient DNA *de novo* assembly validation.

This consists of mainly two steps:
- Ancient DNA damage estimation with [PyDamage](https://github.com/maxibor/pydamage)
- Calling the consensus of the reads aligned back on the contigs (Freebayes + bcftools) to avoid introducing errors due to DNA damage

The proposed way of activating this aDNA subworkflow is through a specific profile `-profile ancient_dna`, as a few general parameters of MAG need to be tuned to make them sensible for aDNA data (the bowtie2 alignment mode).

Would be cool when you have time if you could add the ancient DNA subworfklow to the general MAG workflow schema (on the right of the assembly step) 🙂 .

## To Do

- [x] Make it work with all assemblers
- [ ] Make the bowtie2 alignment process accept `options.args`

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`) (won't lint because of outdated multiqc config @skrakau @d4straub )
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -148,6 +148,8 @@ If `-profile` is not specified, the pipeline will run locally and expect all sof
 * `test`, `test_hybrid`, `test_host_rm`, `test_hybrid_host_rm`, `test_busco_auto`
     * Profiles with a complete configuration for automated testing
     * Includes links to test data so needs no other parameters
+* `ancient_dna`
+    * A profile to run MAG with ancient DNA assembly specific parameters and the ancient DNA subworkflow.
```
Comment (d4straub) [inline docs/usage.md:152]:
```suggestion
    * A profile to run nf-core/mag with ancient DNA assembly specific parameters and the ancient DNA subworkflow.
```

File: workflows/mag.nf [hunk 2]
```diff
@@ -568,6 +568,18 @@ workflow MAG {
...
     }
 
+    /*
+    ================================================================================
+                                    Ancient DNA
+    ================================================================================
+    */
+
+    if (params.ancient_dna){
+        METABAT2_BINNING.out.assembly_mappings.view()
```
Comment (d4straub) [inline workflows/mag.nf:578]: That isnt needed, is it?

**Review by d4straub — COMMENTED** [review]
Nice work!
The remaining test failures seem only linting and master/dev PR confusion.

Just to make sure I understand, this only evaluates whether contigs are likely of ancient origin? Wouldn't it be good when there is some kind of consequence to it, e.g. that those that are of no ancient origin are filtered out, or that the ones or ancient origin are separated (sorry for my ignorance, I have no idea how this usually works).

I am not sure whether the `-profile` is the best place to add those configurations, why do you prefer that over a parameter which is than documented in `nextflow_schema.json`? The documentation of the profile seems not clear this way. In any case, I'd expect aDNA also to be described in the `usage.md` at a more prominent place, such as before `Running the pipeline`.

File: workflows/mag.nf (hunk 2 above)
Comment (maxibor) [reply workflows/mag.nf:599]: Indeed, that was just for debugging ;)

**Comment by maxibor** [comment]
> Nice work! The remaining test failures seem only linting and master/dev PR confusion.
> 
> Just to make sure I understand, this only evaluates whether contigs are likely of ancient origin? Wouldn't it be good when there is some kind of consequence to it, e.g. that those that are of no ancient origin are filtered out, or that the ones or ancient origin are separated (sorry for my ignorance, I have no idea how this usually works).
> 
> I am not sure whether the `-profile` is the best place to add those configurations, why do you prefer that over a parameter which is than documented in `nextflow_schema.json`? The documentation of the profile seems not clear this way. In any case, I'd expect aDNA also to be described in the `usage.md` at a more prominent place, such as before `Running the pipeline`.

I suggest to use the `-profile ancient_dna`, because on top of the ancient_dna subworkflow, ancient DNA alignment needs to have specific parameters (in bowtie2).

File: docs/output.md [hunk 3]
```diff
@@ -400,6 +400,38 @@ If the parameters `--cat_db_generate` and `--save_cat_db` are set, additionally
...
+
+### `PyDamage`
+
+[Pydamage](https://github.com/maxibor/pydamage), is tool to automate the process of ancient DNA damage identification and estimation
+
+<details markdown="1">
+<summary>Output files</summary>
+
+* `Ancient_DNA/pydamage/analyze`
+    * `[sample/group]/pydamage_results/pydamage_results.csv`: PyDamage raw result tabular file in `.csv` format. Format described here: [pydamage.readthedocs.io/en/latest/output.html](https://pydamage.readthedocs.io/en/latest/output.html)
```
[the remark this answers is not in the document]
Comment (maxibor) [reply docs/output.md]: Changed

**Comment by d4straub** [comment]
> > Nice work! The remaining test failures seem only linting and master/dev PR confusion.
> > Just to make sure I understand, this only evaluates whether contigs are likely of ancient origin? Wouldn't it be good when there is some kind of consequence to it, e.g. that those that are of no ancient origin are filtered out, or that the ones or ancient origin are separated (sorry for my ignorance, I have no idea how this usually works).
> > I am not sure whether the `-profile` is the best place to add those configurations, why do you prefer that over a parameter which is than documented in `nextflow_schema.json`? The documentation of the profile seems not clear this way. In any case, I'd expect aDNA also to be described in the `usage.md` at a more prominent place, such as before `Running the pipeline`.
> 
> I suggest to use the `-profile ancient_dna`, because on top of the ancient_dna subworkflow, ancient DNA alignment needs to have specific parameters (in bowtie2).

Do you mean `bowtie2_mode = "--very-sensitive -N 1"`? Why does that need to be in the profile? Adjusting `params.bowtie2_mode` in the workflow conditional on `params.ancient_dna` wouldn't work (e.g. `if (params.ancient_dna) { params.bowtie2_mode  = "--very-sensitive -N 1"}`)? 
I just do not get (yet?) why that needs to be done in a profile. I am not strictly against a profile for that case, but I do not really see its benefit over some adjustments to the defaults somewhere in the `mag.nf` (or any other) either.

**Comment by maxibor** [comment]
> > > Nice work! The remaining test failures seem only linting and master/dev PR confusion.
> > > Just to make sure I understand, this only evaluates whether contigs are likely of ancient origin? Wouldn't it be good when there is some kind of consequence to it, e.g. that those that are of no ancient origin are filtered out, or that the ones or ancient origin are separated (sorry for my ignorance, I have no idea how this usually works).
> > > I am not sure whether the `-profile` is the best place to add those configurations, why do you prefer that over a parameter which is than documented in `nextflow_schema.json`? The documentation of the profile seems not clear this way. In any case, I'd expect aDNA also to be described in the `usage.md` at a more prominent place, such as before `Running the pipeline`.
> > 
> > 
> > I suggest to use the `-profile ancient_dna`, because on top of the ancient_dna subworkflow, ancient DNA alignment needs to have specific parameters (in bowtie2).
> 
> Do you mean `bowtie2_mode = "--very-sensitive -N 1"`? Why does that need to be in the profile? Adjusting `params.bowtie2_mode` in the workflow conditional on `params.ancient_dna` wouldn't work (e.g. `if (params.ancient_dna) { params.bowtie2_mode = "--very-sensitive -N 1"}`)? I just do not get (yet?) why that needs to be done in a profile. I am not strictly against a profile for that case, but I do not really see its benefit over some adjustments to the defaults somewhere in the `mag.nf` (or any other) either.

Could also do like this :)

**Comment by maxibor** [comment]
Hey @d4straub and @jfy133 ,
This PR is now ready for (re)review.
The only issue left is that bowtie2 options are not properly passed to the process, even though it looks to me it's properly set up:

In [conf/modules.config#L99](https://github.com/maxibor/mag/blob/467e2699791b90fa99587d3379ed6491d103537a/conf/modules.config#L99)
```groovy
'bowtie2_assembly_align' {
            args            = params.bowtie2_mode ? params.bowtie2_mode : params.ancient_dna ? '--very-sensitive-local -N 1' : ''
            publish_files   = ['log':'']
            publish_by_meta = ['assembler', 'QC', 'id']
            publish_dir     = "Assembly"
        }
```

With `params.ancient_dna = true`, the bowtie2 command still looks like this:


```bash
$ cat .command.sh
#!/bin/bash -euo pipefail
INDEX=`find -L ./ -name "*.rev.1.bt2l" -o -name "*.rev.1.bt2" | sed 's/.rev.1.bt2l//' | sed 's/.rev.1.bt2//'`
bowtie2 \
    -p "2" \
    -x $INDEX \
     \ # here should be --very-sensitive-local -N 1
    -1 "test_minigut.phix_removed.unmapped_1.fastq.gz" -2 "test_minigut.phix_removed.unmapped_2.fastq.gz" \
    2> "SPAdes-test_minigut-test_minigut.bowtie2.log" |
    samtools view -@ "2" -bS |         samtools sort -@ "2" -o "SPAdes-test_minigut-test_minigut.bam"
samtools index "SPAdes-test_minigut-test_minigut.bam"

if [ SPAdes-test_minigut-test_minigut = "SPAdes-test_minigut-test_minigut" ] ; then
    mv "SPAdes-test_minigut-test_minigut.bowtie2.log" "SPAdes-test_minigut.bowtie2.log"
fi

echo $(bowtie2 --version 2>&1) | sed 's/^.*bowtie2-align-s version //; s/ .*$//' > bowtie2_assembly.version.txt
```

My guess is that a switch to the nf-core modules DSL2 v2 syntax should fix it (like for the aDNA modules of this PR)

File: docs/usage.md [hunk 4]
```diff
@@ -66,6 +66,10 @@ Please note the following requirements:
 
 Again, by default, the group information is only used to compute co-abundances for the binning step, but not for group-wise co-assembly (see the parameter docs for [`--coassemble_group`](https://nf-co.re/mag/parameters#coassemble_group) and [`--binning_map_mode`](https://nf-co.re/mag/parameters#binning_map_mode) for more information about how this group information can be used).
 
+### Ancient DNA mode
+
+Using the ancient DNA profile `-profile ancient_dna` (see [`profile` section](https://nf-co.re/mag/2.1.0/usage#profile)), you can perform ancient DNA *de novo* assembly validation.
```
Comment (jfy133) [inline docs/usage.md:71]: Maybe woth expanding a bit more what it does here (if you switch from profile to parameter)

```suggestion
Using the ancient DNA profile `-profile ancient_dna` (see [`profile` section](https://nf-co.re/mag/2.1.0/usage#profile)), you can perform ancient DNA *de novo* assembly validation.
```

File: nextflow_schema.json [hunk 5]
```diff
@@ -548,6 +548,11 @@
                     "description": "Maximal number of contigs that are not part of any bin but treated as individual genome.",
                     "help_text": "Contigs that do not fulfill the thresholds of `--min_length_unbinned_contigs` and `--max_unbinned_contigs` are pooled for downstream analysis and reporting, except contigs that also do not fullfill `--min_contig_size` are not considered further."
                 },
+                "bowtie2_mode": {
+                    "type": "string",
+                    "description": "Bowtie2 alignment mode",
+                    "help_text": "Bowtie2 alignment mode options, for example: `--very-fast` , `--very-sensitive-local -N 1` , ..."
```
Comment (jfy133) [inline nextflow_schema.json:554]:
```suggestion
                    "help_text": "Bowtie2 alignment mode options, for example: `--very-fast` , `--very-sensitive-local -N 1`"
```

File: subworkflows/local/ancient_dna.nf [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                                .join(BCFTOOLS_INDEX_POST.out.tbi)
+                                .join(input.map { item -> [ item[0], item[1] ] }))
+
+        ch_versions = Channel.empty()
+        ch_versions = PYDAMAGE_ANALYZE.out.versions.first()
+        ch_versions = ch_versions.mix(FAIDX.out.versions.first())
+        ch_versions = ch_versions.mix(FREEBAYES.out.versions.first())
+        ch_versions = ch_versions.mix(BCFTOOLS_CONSENSUS.out.versions.first())
+    emit:
+        contigs_recalled          = BCFTOOLS_CONSENSUS.out.fasta // channel: [ val(meta), path(fasta) ]
```
Comment (jfy133) [inline subworkflows/local/ancient_dna.nf:39]: Would it make sense to also spit out the VCF file if you need to check more closely why a consensus looks iffy?

File: workflows/mag.nf [hunk 7]
```diff
@@ -487,25 +488,56 @@ workflow MAG {
...
-            ch_assemblies,
-            ch_short_reads
-        )
+        if (params.ancient_dna) {
+            METABAT2_BINNING (
+                BINNING_PREPARATION.out.grouped_mappings
+                    .join(ANCIENT_DNA_ASSEMLY_VALIDATION.out.contigs_recalled)
+                    .map{ it -> [ it[0], it[4], it[2], it[3] ] }, // [meta, contigs_recalled, bam, bais]
+                ch_short_reads
+            )
```
Comment (jfy133) [inline workflows/mag.nf:532]: For readability reasons I don't particularly like nesting channel manipulation within module declarations, but rather prepare the channel outside the module declaration, and supply just that (but that's just me)

File: docs/usage.md (hunk 4 above)
Comment (maxibor) [reply docs/usage.md]: This is outdated, sorry, there is no `aDNA` profile anymore

File: nextflow_schema.json [hunk 8]
```diff
@@ -584,6 +589,33 @@
                     "help_text": "Useful to allow reproducibility, as BUSCO datasets are frequently updated and old versions do not always remain accessible."
                 }
             }
+        },
+        "ancient_dna_assembly": {
+            "title": "Ancient DNA assembly",
+            "type": "object",
+            "description": "",
```
Comment (jfy133) [inline nextflow_schema.json:596]: Short description what it does here would be nice

**Review by jfy133 — COMMENTED** [review]
It would still be good to reference that aDNA mode still works in the USAGE docs.

Might as well also mention in in the main README that this is now possible with MAG

But otherwise the code looks good!

File: workflows/mag.nf (hunk 7 above)
Comment (maxibor) [reply workflows/mag.nf:532]: I'll leave as is for now, as this is what allows to pipe into the next step more easily.

File: subworkflows/local/ancient_dna.nf (hunk 6 above)
Comment (maxibor) [reply subworkflows/local/ancient_dna.nf:39]: It's already save in PublishDir (from `modules.config`)

File: nextflow_schema.json (hunk 8 above)
Comment (maxibor) [reply nextflow_schema.json:596]: Done (Added: `Performs ancient DNA assembly validation and contig consensus sequence recalling.`)

File: nextflow_schema.json (hunk 5 above)
Comment (maxibor) [reply nextflow_schema.json:554]: Why ? The ellipsis is there to show that the user can add bowtie2 parameter of their choice

**Comment by maxibor** [comment]
> It would still be good to reference that aDNA mode still works in the USAGE docs.
> 
> Might as well also mention in in the main README that this is now possible with MAG
> 
> But otherwise the code looks good!

I agree, would be nice to update the 

README, however, I suck at Inkscape, @d4straub Do you want to add it ? 

That's what I did

mag_adna

**Comment by jfy133** [comment]
@maxibor I can also do it if you send me the SVG

File: conf/test_ancient_dna.config [hunk 9]
```diff
@@ -0,0 +1,36 @@
+/*
+========================================================================================
+    Nextflow config file for running minimal tests
+========================================================================================
+    Defines input files and everything required to run a fast and simple pipeline test.
+
+    Use as follows:
+        nextflow run nf-core/mag -profile test,<docker/singularity>
```
Comment (d4straub) [inline conf/test_ancient_dna.config:8]:
```suggestion
        nextflow run nf-core/mag -profile test_ancient_dna,<docker/singularity>
```

**Review by d4straub — COMMENTED** [review]
Looks quite good to me.

However, I think it would be really good to mention it in README (at least a small note about aDNA) and docs/usage.
I also agree that it should be in the overview figure (but I disagree to stretch the figure [i.e. rather minimize white space] and it should be obvious that its an on-demand or at least optional step, imho).

Edit: Just saw in the failed linting (expected, I know) the unexpected line
`freebayes | Local copy of module does not match remote`
thats a bit worrisome if its a relevant change. Should be local then.

File: conf/test_ancient_dna.config (hunk 9 above)
Comment (maxibor) [reply conf/test_ancient_dna.config]: Done :)

**Comment by maxibor** [comment]
> Edit: Just saw in the failed linting (expected, I know) the unexpected line
freebayes | Local copy of module does not match remote
thats a bit worrisome if its a relevant change. Should be local then.

That was just one `nf-core modules update freebayes` away, now it's updated 😉

File: README.md [hunk 10]
```diff
@@ -37,6 +37,7 @@ The pipeline then:
 * predicts protein-coding genes for the assemblies using [Prodigal](https://github.com/hyattpd/Prodigal)
 * performs metagenome binning using [MetaBAT2](https://bitbucket.org/berkeleylab/metabat/src/master/), and checks the quality of the genome bins using [Busco](https://busco.ezlab.org/)
 * assigns taxonomy to bins using [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk) and/or [CAT](https://github.com/dutilh/CAT)
+* performs (optionaly) ancient DNA assembly validation using [PyDamage](https://github.com/maxibor/pydamage) and contig consensus sequence recalling with [Freebayes](https://github.com/freebayes/freebayes) and [BCFtools](http://samtools.github.io/bcftools/bcftools.html).
```
Comment (d4straub) [inline README.md:40]: I'd rather put that after assembly, similar to the figure

Comment (maxibor) [reply README.md]: Done

File: README.md [hunk 11]
```diff
@@ -34,6 +34,7 @@ The pipeline then:
 
 * assigns taxonomy to reads using [Centrifuge](https://ccb.jhu.edu/software/centrifuge/) and/or [Kraken2](https://github.com/DerrickWood/kraken2/wiki)
 * performs assembly using [MEGAHIT](https://github.com/voutcn/megahit) and [SPAdes](http://cab.spbu.ru/software/spades/), and checks their quality using [Quast](http://quast.sourceforge.net/quast)
+* performs (optionaly) ancient DNA assembly validation using [PyDamage](https://github.com/maxibor/pydamage) and contig consensus sequence recalling with [Freebayes](https://github.com/freebayes/freebayes) and [BCFtools](http://samtools.github.io/bcftools/bcftools.html)
```
Comment (jfy133) [inline README.md:37]:
```suggestion
* (optionally) performs ancient DNA assembly validation using [PyDamage](https://github.com/maxibor/pydamage) and contig consensus sequence recalling with [Freebayes](https://github.com/freebayes/freebayes) and [BCFtools](http://samtools.github.io/bcftools/bcftools.html)
```

File: nextflow_schema.json (hunk 5 above)
Comment (jfy133) [reply nextflow_schema.json:554]: Ah sorry, I missed `for example` 😅, I thought these were actual options. Ignore!

**Review by jfy133 — APPROVED** [review]
Only think I find missing would be a short description in USAGE of exactly what the purposes of the aDNA workflow is (the README describes what, not why), but otherwise wise looks good :+1:

**Review by d4straub — APPROVED** [review]
Looks good to me as well. I agree and still think (previous comments) that there should be a mention of the profile and usage in the docs, i.e. `usage.md`. I approve it anyway.

**Pull request merged**